lib/gems/pending/util/object_storage/miq_swift_storage.rb

```diff
@@ -0,0 +1,155 @@
+require 'pp'
+require 'util/miq_object_storage'
+
+class MiqSwiftStorage < MiqObjectStorage
+  attr_reader :container_name
+
+  def self.uri_scheme
+    "swift".freeze
+  end
+
+  def self.new_with_opts(opts)
+    new(opts.slice(:uri, :username, :password))
+  end
+
+  def initialize(settings)
+    super(settings)
+
+    # NOTE: This line to be removed once manageiq-ui-class region change implemented.
+    @bucket_name         = URI(@settings[:uri]).host
+
+    raise "username and password are required values!" if @settings[:username].nil? || @settings[:password].nil?
+    _scheme, _userinfo, @host, @port, _registry, @mount_path, _opaque, query, _fragment = URI.split(URI.encode(@settings[:uri]))
+    query_params(query) if query
+    @swift          = nil
+    @username       = @settings[:username]
+    @password       = @settings[:password]
+    @container_name = @mount_path[0] == File::Separator ? @mount_path[1..-1] : @mount_path
+  end
+
+  def uri_to_local_path(remote_file)
+    # Strip off the leading "swift:/" from the URI"
+    File.join(@mnt_point, URI(remote_file).host, URI(remote_file).path)
+  end
+
+  def uri_to_object_path(remote_file)
+    # Strip off the leading "swift://" and the container name from the URI"
+    # Also remove the leading delimiter.
+    object_file_with_bucket = URI.split(URI.encode(remote_file))[5]
+    object_file_with_bucket.split(File::Separator)[2..-1].join(File::Separator)
+  end
+
+  def upload_single(dest_uri)
+    #
+    # Get the remote path, and parse out the bucket name.
+    #
+    object_file = uri_to_object_path(dest_uri)
+    #
+    # write dump file to swift
+    #
+    logger.debug("Writing [#{source_input}] to => Bucket [#{container_name}] using object file name [#{object_file}]")
+    begin
+      swift_file = container.files.new(:key => object_file)
+      params     = {
+        :expects       => [201, 202],
+        :headers       => {},
+        :request_block => -> { read_single_chunk },
+        :idempotent    => false,
+        :method        => "PUT",
+        :path          => "#{Fog::OpenStack.escape(swift_file.directory.key)}/#{Fog::OpenStack.escape(swift_file.key)}"
+      }
+      swift_file.service.send(:request, params)
```

Comment:
Why are we using `send` here? Is there not a public method for what we want to do?

Reply:
Pretty sure it is, but let me double check.

* * *

Normally, we would want to do `#put_object`, but that has an issue with the `:expects` that I am defining here since it currently only supports `201` responses:

https://github.com/fog/fog-openstack/blob/c4ccd3319726c935b406eaf32cf339999fa89380/lib/fog/storage/openstack/requests/put_object.rb#L18-L38

But that isn't what gets returned when doing a streamed upload, which is what is being done here.  Also, the streaming part isn't exposed by `fog/openstack`, as explained here:

https://github.com/ManageIQ/manageiq-gems-pending/commit/56ba9e7c5caef5f76a12055ab3871f0729b878a8

Reply:
> Pretty sure it is, but let me double check

Okay, yeah.  Got a little confused where the method was coming from (thought it was `fog-core`, but I think the `fog/openstack` folks are doing a few things that make the method private:

https://github.com/fog/fog-openstack/blob/c4ccd3319726c935b406eaf32cf339999fa89380/lib/fog/openstack/core.rb#L76-L81

And I did test in a console, and that method is indeed private there as well:

```ruby
irb> require 'fog/openstack'
irb> client = Fog::Storage::Openstack.new(your_params_here)
irb> client.request
#=> NoMethodError: private method `request' called for #<Fog::Storage::OpenStack::Real:0x00007f8adb116c88>
irb> client.method(:request).source_location
#=> ["~/.gem/ruby/2.4.2/gems/fog-openstack-0.1.27/lib/fog/openstack/core.rb", 78]
```

Reply:
Okay, can we add a comment here about why we are not using the public interface then?

Reply:
Done.